File: packages/core/tests/location-hash.test.ts
import {
  composeUnforgettableLocationHash,
  parseUnforgettableLocationHash,
  RecoveryFactor,
} from '../src'

describe('url location hash utils', () => {
  describe('composeUnforgettableLocationHash', () => {
    it('creates valid hash string with all params', () => {
      const params = {
        dataTransferId: 'test-id',
        encryptionPublicKey: 'test-pk',
        factors: [RecoveryFactor.Face, RecoveryFactor.Image, RecoveryFactor.Password],
        walletAddress: '0xabc',
        group: 'm',
        customParams: {
          t: 'test-theme',
          d: 'test-data',
        },
      }

      const hash = composeUnforgettableLocationHash(params)

      expect(hash).toBe('#id=test-id&epk=test-pk&f=1%2C2%2C3&wa=0xabc&g=m&t=test-theme&d=test-data')
    })

    it('it creates a hash which can be parsed into the initial params object', () => {
      const params = {
        dataTransferId: 'test-id',
        encryptionPublicKey: 'test-pk',
        factors: [RecoveryFactor.Face, RecoveryFactor.Image, RecoveryFactor.Password],
        walletAddress: '0xabc',
        group: 'm',
        customParams: {
          t: 'test-theme',
          d: 'test-data',
        },
      }

      const hash = composeUnforgettableLocationHash(params)
      const parsedParams = parseUnforgettableLocationHash(hash)

      expect(parsedParams).toEqual(params)
    })

    it('excludes wallet address from hash if not passed', () => {
      const params = {
        dataTransferId: 'test-id',
        encryptionPublicKey: 'test-pk',
        factors: [RecoveryFactor.Face, RecoveryFactor.Image, RecoveryFactor.Password],
      }

      const hash = composeUnforgettableLocationHash(params)
      const urlSearchParams = new URLSearchParams(hash.slice(1))

      expect(urlSearchParams.get('id')).toBe(params.dataTransferId)
      expect(urlSearchParams.get('epk')).toBe(params.encryptionPublicKey)
      expect(urlSearchParams.get('f')).toBe(params.factors.join(','))
      expect(urlSearchParams.get('wa')).toBeNull()
      expect(hash).toBe('#id=test-id&epk=test-pk&f=1%2C2%2C3')
    })

    it('excludes group from hash if not passed', () => {
      const params = {
        dataTransferId: 'test-id',
        encryptionPublicKey: 'test-pk',
        factors: [RecoveryFactor.Face, RecoveryFactor.Image, RecoveryFactor.Password],
      }

      const hash = composeUnforgettableLocationHash(params)
      const urlSearchParams = new URLSearchParams(hash.slice(1))

      expect(urlSearchParams.get('id')).toBe(params.dataTransferId)
      expect(urlSearchParams.get('epk')).toBe(params.encryptionPublicKey)
      expect(urlSearchParams.get('f')).toBe(params.factors.join(','))
      expect(urlSearchParams.get('g')).toBeNull()
      expect(hash).toBe('#id=test-id&epk=test-pk&f=1%2C2%2C3')
    })
  })

  describe('parseUnforgettableLocationHash', () => {
    describe('location hash parsing', () => {
      it('parses location hash into a valid params object', () => {
        const parsedParams = parseUnforgettableLocationHash(
          '#id=test-id&epk=test-pk&f=1,2,3&wa=0xabc&g=m',
        )

        expect(parsedParams).toEqual({
          dataTransferId: 'test-id',
          encryptionPublicKey: 'test-pk',
          factors: [1, 2, 3],
          walletAddress: '0xabc',
          group: 'm',
        })
      })

      it('parses without #', () => {
        const parsedParams = parseUnforgettableLocationHash(
          'id=test-id&epk=test-pk&f=1,2,3&wa=0xabc&g=m',
        )

        expect(parsedParams).toEqual({
          dataTransferId: 'test-id',
          encryptionPublicKey: 'test-pk',
          factors: [1, 2, 3],
          walletAddress: '0xabc',
          group: 'm',
        })
      })

      describe('parsing custom params', () => {
        it('parses location hash with custom params into a valid params object', () => {
          const parsedParams = parseUnforgettableLocationHash(
            '#id=test-id&epk=test-pk&f=1,2,3&t=light&t=test-theme&d=test-data&r',
          )

          expect(parsedParams).toEqual({
            dataTransferId: 'test-id',
            encryptionPublicKey: 'test-pk',
            factors: [1, 2, 3],
            customParams: {
              t: 'test-theme',
              d: 'test-data',
              r: '',
            },
          })
        })

        it('parses location hash with custom params and does not overwrite a required path params', () => {
          const parsedParams = parseUnforgettableLocationHash(
            '#id=test-id&epk=original-pk&f=1,2,3&epk=overwritten-pk',
          )

          expect(parsedParams).toEqual({
            dataTransferId: 'test-id',
            encryptionPublicKey: 'original-pk',
            factors: [1, 2, 3],
          })
        })
      })
    })

    it('throws an error if the data transfer ID is missing', () => {
      expect(() => parseUnforgettableLocationHash('#epk=test-pk&f=1,2,3&wa=0xabc')).toThrow(
        'Invalid recovery path parameters',
      )
    })

    it('throws an error if the encrypted public key is missing', () => {
      expect(() => parseUnforgettableLocationHash('#id=test-id&f=1,2,3&wa=0xabc')).toThrow(
        'Invalid recovery path parameters',
      )
    })

    describe('parseFactors', () => {
      it('returns empty array if factors are not provided', () => {
        const parsedParams = parseUnforgettableLocationHash('#id=test-id&epk=test-pk')

        expect(parsedParams.factors).toEqual([])
      })

      it('returns empty array if factors key is provided but empty', () => {
        const parsedParams = parseUnforgettableLocationHash('#id=test-id&epk=test-pk&f=')

        expect(parsedParams.factors).toEqual([])
      })

      it('returns the original order in case of repeated factors', () => {
        const parsedParams = parseUnforgettableLocationHash('#id=test-id&epk=test-pk&f=1,2,3,2,1,3')

        expect(parsedParams.factors).toEqual([1, 2, 3, 2, 1, 3])
      })

      it('ignores empty items in the enumeration', () => {
        const parsedParams = parseUnforgettableLocationHash('#id=test-id&epk=test-pk&f=1,,2,3')

        expect(parsedParams.factors).toEqual([1, 2, 3])
      })

      it('throws an error if the factor is not an integer number', () => {
        const invalidFactor = 'abc'

        expect(() =>
          parseUnforgettableLocationHash(`#id=test-id&epk=test-pk&f=1,${invalidFactor},3`),
        ).toThrow(`Recovery factor id is not an integer: "${invalidFactor}"`)
      })

      it('throws an error if the factor is out of the factors registry range', () => {
        const invalidFactor = Number.MAX_SAFE_INTEGER

        expect(() =>
          parseUnforgettableLocationHash(`#id=test-id&epk=test-pk&f=1,2,${invalidFactor}`),
        ).toThrow(`Invalid recovery factor id: ${invalidFactor}`)
      })
    })
  })
})

File: packages/android/src/test/kotlin/com/rarimo/unforgettable/LocationHashTest.kt
package com.rarimo.unforgettable

import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class LocationHashTest {
    
    @Test
    fun `test compose location hash`() {
        val params = UnforgettablePathParams(
            dataTransferId = "test-id-123",
            encryptionPublicKey = "test-key",
            factors = listOf(RecoveryFactor.FACE, RecoveryFactor.PASSWORD),
            walletAddress = "0x123"
        )
        
        val hash = composeUnforgettableLocationHash(params)
        
        assertTrue(hash.startsWith("#"))
        assertTrue(hash.contains("id=test-id-123"))
        assertTrue(hash.contains("epk=test-key"))
        assertTrue(hash.contains("f=1%2C3") || hash.contains("f=1,3"))
        assertTrue(hash.contains("wa=0x123"))
    }
    
    @Test
    fun `test compose location hash without wallet`() {
        val params = UnforgettablePathParams(
            dataTransferId = "test-id",
            encryptionPublicKey = "key",
            factors = listOf(RecoveryFactor.IMAGE)
        )
        
        val hash = composeUnforgettableLocationHash(params)
        
        assertTrue(hash.startsWith("#"))
        assertTrue(hash.contains("id=test-id"))
        assertTrue(hash.contains("epk=key"))
        assertTrue(hash.contains("f=2"))
        assertFalse(hash.contains("wa="))
    }
    
    @Test
    fun `test parse location hash`() {
        val hash = "#id=test-123&epk=my-key&f=1,3&wa=0xabc"
        
        val params = parseUnforgettableLocationHash(hash)
        
        assertEquals("test-123", params.dataTransferId)
        assertEquals("my-key", params.encryptionPublicKey)
        assertEquals(2, params.factors.size)
        assertEquals(RecoveryFactor.FACE, params.factors[0])
        assertEquals(RecoveryFactor.PASSWORD, params.factors[1])
        assertEquals("0xabc", params.walletAddress)
    }
    
    @Test
    fun `test parse location hash without hash prefix`() {
        val hash = "id=test&epk=key&f=1"
        
        val params = parseUnforgettableLocationHash(hash)
        
        assertEquals("test", params.dataTransferId)
        assertEquals("key", params.encryptionPublicKey)
        assertEquals(1, params.factors.size)
        assertEquals(RecoveryFactor.FACE, params.factors[0])
    }
    
    @Test
    fun `test parse invalid location hash throws`() {
        val invalidHash = "#invalid"
        
        assertFailsWith<LocationHashError.InvalidParameters> {
            parseUnforgettableLocationHash(invalidHash)
        }
    }
    
    @Test
    fun `test parse invalid factor throws`() {
        val hash = "#id=test&epk=key&f=99"
        
        assertFailsWith<LocationHashError.InvalidFactorId> {
            parseUnforgettableLocationHash(hash)
        }
    }
    
    @Test
    fun `test round trip location hash`() {
        val originalParams = UnforgettablePathParams(
            dataTransferId = "uuid-123",
            encryptionPublicKey = "public-key-data",
            factors = listOf(RecoveryFactor.FACE, RecoveryFactor.IMAGE, RecoveryFactor.PASSWORD),
            walletAddress = "0x456"
        )
        
        val hash = composeUnforgettableLocationHash(originalParams)
        val parsedParams = parseUnforgettableLocationHash(hash)
        
        assertEquals(originalParams.dataTransferId, parsedParams.dataTransferId)
        assertEquals(originalParams.encryptionPublicKey, parsedParams.encryptionPublicKey)
        assertEquals(originalParams.factors, parsedParams.factors)
        assertEquals(originalParams.walletAddress, parsedParams.walletAddress)
    }
    
    @Test
    fun `test compose location hash with group`() {
        val params = UnforgettablePathParams(
            dataTransferId = "test-id",
            encryptionPublicKey = "key",
            factors = listOf(RecoveryFactor.FACE),
            group = "test-group"
        )
        
        val hash = composeUnforgettableLocationHash(params)
        
        assertTrue(hash.contains("g=test-group"))
    }
    
    @Test
    fun `test parse location hash with group`() {
        val hash = "#id=test&epk=key&f=1&g=my-group"
        
        val params = parseUnforgettableLocationHash(hash)
        
        assertEquals("my-group", params.group)
    }
    
    @Test
    fun `test compose location hash with custom params`() {
        val params = UnforgettablePathParams(
            dataTransferId = "test-id",
            encryptionPublicKey = "key",
            factors = listOf(RecoveryFactor.FACE),
            customParams = mapOf("t" to "dark", "d" to "data")
        )
        
        val hash = composeUnforgettableLocationHash(params)
        
        assertTrue(hash.contains("t=dark"))
        assertTrue(hash.contains("d=data"))
    }
    
    @Test
    fun `test parse location hash with custom params`() {
        val hash = "#id=test&epk=key&f=1&t=light&d=test-data"
        
        val params = parseUnforgettableLocationHash(hash)
        
        assertEquals("light", params.customParams?.get("t"))
        assertEquals("test-data", params.customParams?.get("d"))
    }
    
    @Test
    fun `test parse location hash empty factors`() {
        val hash = "#id=test&epk=key&f="
        
        val params = parseUnforgettableLocationHash(hash)
        
        assertEquals(0, params.factors.size)
    }
    
    @Test
    fun `test parse location hash ignores empty factor items`() {
        val hash = "#id=test&epk=key&f=1,,2,,3"
        
        val params = parseUnforgettableLocationHash(hash)
        
        assertEquals(3, params.factors.size)
        assertEquals(listOf(RecoveryFactor.FACE, RecoveryFactor.IMAGE, RecoveryFactor.PASSWORD), params.factors)
    }
    
    @Test
    fun `test round trip with all params`() {
        val originalParams = UnforgettablePathParams(
            dataTransferId = "uuid-123",
            encryptionPublicKey = "public-key-data",
            factors = listOf(RecoveryFactor.FACE, RecoveryFactor.IMAGE, RecoveryFactor.PASSWORD),
            walletAddress = "0x456",
            group = "test-group",
            customParams = mapOf("t" to "dark", "d" to "data")
        )
        
        val hash = composeUnforgettableLocationHash(originalParams)
        val parsedParams = parseUnforgettableLocationHash(hash)
        
        assertEquals(originalParams.dataTransferId, parsedParams.dataTransferId)
        assertEquals(originalParams.encryptionPublicKey, parsedParams.encryptionPublicKey)
        assertEquals(originalParams.factors, parsedParams.factors)
        assertEquals(originalParams.walletAddress, parsedParams.walletAddress)
        assertEquals(originalParams.group, parsedParams.group)
        assertEquals("dark", parsedParams.customParams?.get("t"))
        assertEquals("data", parsedParams.customParams?.get("d"))
    }
}
